### ISSUE TITLE
Incorrect QuantifiedConstraints parenthesization in type classes

### ISSUE THREAD

**Issue by RyanGlScott** [haskell/haddock#1063]
(Originally noticed [here](http://hackage.haskell.org/package/simplistic-generics-0.1.0.0/docs/Data-Constraints.html#t:Implies).)

Take the following code:

```hs
{-# LANGUAGE ConstraintKinds #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE QuantifiedConstraints #-}
{-# LANGUAGE UndecidableInstances #-}
module Bug where

class    (c => d) => Implies c d
instance (c => d) => Implies c d
```

And run `haddock Bug.hs --html` on it. Trying this with a Haddock built from the `ghc-head` branch (commit a31b562f3f3a33ae2b08f9173ea08506b357f3a5) yields this:

![haddock](https://user-images.githubusercontent.com/2364661/57883853-619eca00-77f5-11e9-8311-1187ec4dee25.png)

This is missing the required parentheses around `c => d` in both the class and instance declarations.

**Comment by harpocrates** [comment]
@RyanGlScott Are you already fixing this? If not, I can get a patch up shortly.

In any case, thanks for the bug report (and for always including screenshots and test cases ❤️)!

**Comment by RyanGlScott** [comment]
Well, my first inclination was to see how GHC is handling this, because _clearly_ GHC must do this correctly, right? Sadly, the reality much harsher than I expected:

```
λ> ; $([d| class (c => d) => Implies c d; instance (c => d) => Implies c d |])
<interactive>:8:5-74: Splicing declarations
    [d| class (c_a4GK => d_a4GL) => Implies_a4GJ c_a4GK d_a4GL
        
        instance (c_a4GM => d_a4GN) => Implies_a4GJ c_a4GM d_a4GN |]
  ======>
    class c_a4HC => d_a4HD => Implies_a4HB c_a4HC d_a4HD
    instance c_a4HE => d_a4HF => Implies_a4HB c_a4HE d_a4HF
```

Yikes! The bottom part of the `-ddump-splices` output is _also_ incorrect. So it appears that there may be something afoot in GHC as well. (I'm as of yet unclear if this buglet is related to the Haddock one observed above.)

I'll try to diagnose this issue first.

**Comment by harpocrates** [comment]
Your `ddump-splices` bug is separate and GHC's fault. The Haddock output is Haddock's fault. The reason for matching bugs is that Haddock ported the precedence logic for pretty-printing from GHC, so I'm not surprised GHC has this issue too. 😄

**Comment by RyanGlScott** [comment]
Yes, upon further thought, I believe that bug is entirely unrelated to what's going on in GHC, which appears to be a `Convert`-specific (i.e., Template Haskell–specific) issue.

As for what the culprit in Haddock might be, I have a hunch. Here is the function which parenthesizes contexts in the HTML backend:

https://github.com/haskell/haddock/blob/a31b562f3f3a33ae2b08f9173ea08506b357f3a5/haddock-api/src/Haddock/Backends/Xhtml/Decl.hs#L476-L479

In particular, single-predicate contexts go through this auxiliary `ppCtxType` function:

https://github.com/haskell/haddock/blob/a31b562f3f3a33ae2b08f9173ea08506b357f3a5/haddock-api/src/Haddock/Backends/Xhtml/Decl.hs#L1109-L1110

That, in turn, invokes `reparenTypePrec PREC_CTX`. What is the `reparenTypePrec` case for `HsQualTy`? As it turns out, it's this:

https://github.com/haskell/haddock/blob/a31b562f3f3a33ae2b08f9173ea08506b357f3a5/haddock-api/src/Haddock/GhcUtils.hs#L279-L280

It's guarded by `paren p PREC_FUN`, which means that it will only be parenthesized if the supplied `Precedence` is at least as great as `PREC_FUN`. But the definition of `Precendence` shows that `PREC_CTX` is less than `PREC_FUN`, according to its derived `Ord` instance:

https://github.com/haskell/haddock/blob/a31b562f3f3a33ae2b08f9173ea08506b357f3a5/haddock-api/src/Haddock/GhcUtils.hs#L240-L253

-----

What to do? We could just merge `PREC_CTX` and `PREC_FUN`, as I'm unclear on what utility it brings to make a distinction between the two. Disclaimer: I haven't tested this change, so I don't know yet if this would cause any other adverse changes.

**Comment by harpocrates** [comment]
Yep, you found the right place. Perhaps we need another precedence level? I have no time right now, but I'll look again tomorrow morning at this.

**Comment by RyanGlScott** [comment]
FWIW, if I remove the `PREC_CXT` constructor and simply replace all occurrences of it with `PREC_FUN` then the test suite passes save for two tests:

```diff
Diff for file "TypeOperators.html":
144c144
<           > :: (a ~ b) =&gt; a -&gt; b <a href="#" class="selflink"
---
>           > :: a ~ b =&gt; a -&gt; b <a href="#" class="selflink"
160c160
<           > :: ((a <a href="#" title="TypeOperators"
---
>           > :: (a <a href="#" title="TypeOperators"
166c166
<           > a)) =&gt; a <a href="#" class="selflink"
---
>           > a) =&gt; a <a href="#" class="selflink"
Diff for file "FunArgs.html":
221c221
<                 > (b :: ()). (d ~ '<a href="#" title="GHC.Tuple"
---
>                 > (b :: ()). d ~ '<a href="#" title="GHC.Tuple"
223c223
<                 >)</td
---
>                 ></td
```

It looks like that change causes `(a ~ b) => Blah` to be rendered as `a ~ b => Blah` instead. Both are valid GHC, although I don't know if you'd prefer to keep the parentheses in this situation.

**Comment by harpocrates** [comment]
I have no personal preference for `(a ~ b) => Blah` or `a ~ b => Blah`, but it would be nice to have a good reason to change the current behaviour if we are going to do so. I picked the current behaviour based on what GHC does:

```
λ> ; $([d| a :: (Int ~ b) => b -> Int; a = id |])
<interactive>:18:5-45: Splicing declarations
    [d| a_a6ay :: (Int ~ b_a6az) => b_a6az -> Int
        a_a6ay = id |]
  ======>
    a_a6bf :: Int ~ b_a6be => b_a6be -> Int
    a_a6bf = id
```

When trying out these sorts of changes, you may want to rebuild boot lib docs and `diff -rq` the new and old docs, just to make sure nothing big has been broken. The testsuite is great for preventing regressions, but unfortunately still has pretty poor coverage.

I haven't checked, but I suspect the changes you propose might have some pretty large effects.

**Comment by RyanGlScott** [comment]
I don't have a personal preference either. In fact, #1064 looks much less invasive than what I was proposing, so I'd be perfectly content with that.

**Comment by harpocrates** [comment]
Fixed in 384577e862171bdedc9311c9d17f7ad3a4a33456.
